camlog/camlog.1
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139,
.\" USA.
.TH camlog 1
.SH NAME
camlog \- Image stream logger
.SH SYNOPSIS
.TP 5
\fBcamlog \fI[options]\fR

.SH DESCRIPTION
.PP
\fBcamlog\fR is a tool for writing video data to disk, primarily to save
data for post-processing and analysis.  The video source written to
disk is determined by specifying either the -i or -c flag -- exactly
one of these must be specified.  Logfiles produced by camlog are
suitable for loading with the Camunits input.log unit.

Using the -i option is simpler, and is useful for dealing with an
input unit where the default settings are sufficient.  If the default
unit settings are not sufficient, or pre-log processing is desired,
then use the -c option and specify a chain description file as produced
by camview or the cam_unit_chain_snapshot() function.

.SH OPTIONS
The following options are provided by \fBcamlog\fR using the standard GNU
command line syntax:
.TP
.B \-i, \-\-input=\fIID\fB
Loads the specified input source.  \fISOURCE\fR should be a Camunits unit id.
.TP
.B \-c, \-\-chain=\fINAME\fB
Loads the chain from file NAME.
.TP
.B \-o, \-\-output=\fIFILE\fB
Saves result to filename FILE.  If not specified, a filename is automatically
chosen.
.TP
.B \-f, \-\-force
Force overwrite of output_file if it already exists.  If not specified, and the output file already exists, then a suffix is automatically appended to the
filename to prevent overwriting existing files.
.TP
.B \-n, \-\-no-write
Do not write video data to disk.  Useful for testing.
.TP
.B \-v, \-\-verbose
Be more verbose.
.TP
.B \-h, \-\-help
Print this help text and exit.

.SH AUTHORS

 Albert Huang <\fIashuang@alum.mit.edu\fR>
.br
 David Moore <\fIdcm@acm.org\fR>


NEWS
camunits 0.1.1 - Dec 17, 2008
=============================

* camview: 
  - add FPS display, other minor UI tweaks
  - change "-f" option to "-c"
* libcamunits:
  - CamInputLog: add looping controls
  - add CAM_PIXEL_FORMAT_LE_GRAY16, opengl support
  - CamLoggerUnit: add "auto-suffix-enable" control, option to disable
                   auto suffix on filename
  - rename IS_ALIGNEDxx macros to CAM_IS_ALIGNEDxx
  - add better type safety to CamUnitControl
  - CamUnitManager bugfixes
* libcamunits-gtk:
  - bugfixes in UnitControlWidget
* camlog
  - add "-c" option to load chain from XML description file

camunits 0.1.0 - Nov 11, 2008
=============================

NEW PROJECT NAME: camunits.

* better support for 16-bit pixel formats
* input_dc1394: include format7 mode in output format string
* unit_manager: on failure to open plugin dir, warn in dbg instead of stderr
* unit_chain:   store output format name in saved XML.
* unit_control_widget: slightly smarter about format changing
* minor bugfixes all around.

This release breaks the old API.  To port libcam code to Camunits code, 
replace all instances of "libcam" and "LIBCAM" with "camunits" and "CAMUNITS",
respectively.  This applies to source code, include paths, library names,
linker flags, environment variables, pkg-config names, etc.

The following two perl commands accomplish the search/replace:

   $ perl -pi -e "s/LIBCAM/CAMUNITS/g;" <files>
   $ perl -pi -e "s/libcam/camunits/g;" <files>

libcam 0.0.9 - Jul 16, 2008
===========================
* unit manager widget UI enhancements
  - sort unit descriptions by unit name
  - toggle expansion of packages when package is double-clicked/activated
* enable spinbutton for float unit control widget
* warn when units produce framebuffers with zero bytesused and timestamp fields
* fix some docs
* fix accidental API change in 0.0.8.

libcam 0.0.8 - Jun 13, 2008
===========================
* add --no-gui option to camview
* input_example: fix timestamps
* camlog: fix thrashing for inexact log seeking (e.g. timestamps)
* update tutorial to match reality
* add qt4 examples
* input_v4l: was not populating timestamp, bytesused.  fixed.

libcam 0.0.7 - Apr 17, 2008
===========================
* Fix debayer regression from last version's portability improvements
* Allow specification of extra plugin path at configure time

libcam 0.0.6 - Apr 9, 2008
==========================
* Add asynchronous driver handling
* Portability improvements (can build non x86)

libcam 0.0.5 - Feb 14, 2008
===========================
* updates across the board.
* API changed

libcam 0.0.4 - Jan 8, 2008
==========================
* Update DC1394 plugin to use libdc1394-2.0.0
* Improvements to input.log

libcam 0.0.3 - Dec 19, 2007
==========================
* Fix installation directory of libcam-gtk headers and renamed cam_gtk.h to
  cam-gtk.h.

libcam 0.0.2
============
* Now builds under Mac OS X (tested on Leopard)
* Tutorial and examples improvements
* Documentation improvements
* Update DC1394 plugin to use libdc1394-2.0.0-rc9
* Allow DC1394 modes other than Format 7
* Added additional colorspace conversions
* Allow bayer demosaic code to work on non-SSE3 CPUs (still requires SSE2)

libcam 0.0.1
============

* Initial release


trunk/camlog/camlog.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <getopt.h>
#include <unistd.h>

#include <glib.h>

#include <camunits/cam.h>

#include "signal_pipe.h"

typedef struct _state_t {
    int verbose;
    int frameno;
    int64_t lasttime;
} state_t;

#define FRAMES_PER_PRINTF   100

static int64_t _timestamp_now()
{
    struct timeval tv;
    gettimeofday (&tv, NULL);
    return (int64_t) tv.tv_sec * 1000000 + tv.tv_usec;
}

static void
on_frame_ready (CamUnitChain *chain, CamUnit *unit, const CamFrameBuffer *buf, 
        void *user_data)
{
    state_t *self = user_data;

    if (self->verbose) {
        printf("frame: %d\n", self->frameno);
    }
    if (self->frameno % FRAMES_PER_PRINTF == 0) {
        int64_t timestamp;
        timestamp = _timestamp_now ();
        if (self->lasttime != 0) {
            printf ("%d frames at %.1f Hz\n", self->frameno,
                    FRAMES_PER_PRINTF * 1000000.0 /
                    (timestamp - self->lasttime));
        }
        self->lasttime = timestamp;
    }
    self->frameno++;
}

static void
print_inputs (CamUnitManager *manager)
{
    fprintf(stderr, "Units available in package input:\n\n"); 
    GList *udlist = cam_unit_manager_list_package (manager, "input", TRUE);
    for (GList *uditer=udlist; uditer; uditer=uditer->next) {
        CamUnitDescription *udesc = CAM_UNIT_DESCRIPTION(uditer->data);
        printf("  %s  (%s)\n", udesc->unit_id, udesc->name);
    }
    g_list_free(udlist);
}

static void
usage()
{
    fprintf(stderr, 
        "Usage: camlog [OPTIONS]\n"
        "\n"
        "camlog is a tool for writing video data to disk, primarily to save\n"
        "data for post-processing and analysis.  The video source written to\n"
        "disk is determined by specifying either the -i or -c flag -- exactly\n"
        "one of these must be specified.  Logfiles produced by camlog are\n"
        "suitable for loading with the Camunits input.log unit\n"
        "\n"
        "Using the -i option is simpler, and is useful for dealing with an\n"
        "input unit where the default settings are sufficient.  If the default\n"
        "unit settings are not sufficient, or pre-log processing is desired,\n"
        "then use the -c option and specify a chain description file as produced\n"
        "by camview or the cam_unit_chain_snapshot() function.\n"
        "\n"
        "Options:\n"
        " -h, --help         Shows this help text\n"
        " -i, --input ID     Use the specified Camunit ID as input.\n"
        " -c, --chain NAME   Load chain from file NAME.\n"
        "\n"
        " -o, --output FILE  Saves result to filename FILE.  If not specified, a\n"
        "                    filename is automatically chosen.\n"
        " -f, --force        Force overwrite of output_file if it already exists\n"
        "                    If not specified, and the output file already\n"
        "                    exists, then a suffix is automatically appended\n"
        "                    to the filename to prevent overwriting existing\n"
        "                    files.\n"
        " -n, --no-write     Do not write video data to disk.  Useful for testing.\n"
        " -v, --verbose      Print information about each frame.\n\n");
}

int main(int argc, char **argv)
{
    int status = 1;

    char *log_fname = NULL;
    char *input_id = NULL;
    char *chain_fname = NULL;
    int overwrite = 0;
    int do_logging = 1;
    GMainLoop *mainloop = NULL;
    state_t *self = (state_t*)calloc(1, sizeof(state_t));
    self->verbose = 0;
    self->frameno = 0;

    setlinebuf (stdout);
    setlinebuf (stderr);

    char *optstring = "hi:c:o:fnv";
    char c;
    struct option long_opts[] = { 
        { "help", no_argument, 0, 'h' },
        { "input", required_argument, 0, 'i' },
        { "chain", required_argument, 0, 'c' },
        { "output", no_argument, 0, 'o' },
        { "force", no_argument, 0, 'f' },
        { "no-write", no_argument, 0, 'n' },
        { "verbose", no_argument, 0, 'v' },
        { 0, 0, 0, 0 }
    };

    g_type_init();

    while ((c = getopt_long (argc, argv, optstring, long_opts, 0)) >= 0)
    {
        switch (c) {
            case 'i':
                free(input_id);
                input_id = strdup(optarg);
                break;
            case 'c':
                free(chain_fname);
                chain_fname = strdup(optarg);
                break;
            case 'o':
                free(log_fname);
                log_fname = strdup(optarg);
                break;
            case 'f':
                overwrite = 1;
                break;
            case 'n':
                do_logging = 0;
                break;
            case 'v':
                self->verbose = 1;
                break;
            case 'h':
            default:
                usage();
                return 1;
        };
    }

    // setup the image processing chain
    CamUnitChain * chain = cam_unit_chain_new();

    // XXX hack... don't remember why this is there anymore...
    sleep (1);

    if (!input_id && !chain_fname) {
        usage(); 
        print_inputs(chain->manager);
        goto done;
    } else if(input_id && chain_fname) {
        fprintf(stderr, "Only one of -c and -i can be specified\n");
        usage();
        goto done;
    }

    // create the GLib mainloop
    mainloop = g_main_loop_new (NULL, FALSE);
    signal_pipe_glib_quit_on_kill (mainloop);

    // instantiate the input unit
    if(input_id) {
        CamUnit *input = cam_unit_chain_add_unit_by_id(chain, input_id);
        if (! input) {
            fprintf(stderr, "CRAP!  Couldn't create input unit [%s].\n\n", 
                    input_id);
            print_inputs(chain->manager);
            goto done;
        }

        cam_unit_stream_init (input, NULL);
    } else {
        char *xml_str = NULL;
        if(g_file_get_contents(chain_fname, &xml_str, NULL, NULL)) {
            cam_unit_chain_load_from_str(chain, xml_str, NULL);
        }
        free(xml_str);
    }

    CamUnit *logger_unit = NULL;
    if (do_logging) {
        // create the logger unit and add it to the chain.
        logger_unit = cam_unit_chain_add_unit_by_id(chain, 
                "output.logger");
        assert(logger_unit);

        // set the filename, and start the logger unit recording
        if (log_fname) {
            cam_unit_set_control_string(logger_unit, "desired-filename", 
                    log_fname);
            cam_unit_set_control_boolean(logger_unit, "auto-suffix-enable",
                    !overwrite);
        }
        cam_unit_set_control_boolean (logger_unit, "record", TRUE);

        // print the actual filename
        char *fname = g_object_get_data(logger_unit, "actual-filename");
        if(fname) 
            fprintf(stderr, "logging to: %s\n", fname);

    } else {
        fprintf (stderr, "not logging to disk\n");
    }

    // start the chain streaming
    CamUnit *faulty_unit = cam_unit_chain_all_units_stream_init (chain);

    // did everything start up correctly?
    if (faulty_unit) {
        fprintf (stderr, "Unit [%s] is not ready, aborting...\n",
                cam_unit_get_name (faulty_unit));
        goto done;
    }

    cam_unit_chain_attach_glib (chain, 1000, NULL);
    g_signal_connect (G_OBJECT (chain), "frame-ready",
            G_CALLBACK (on_frame_ready), self);

    // run the main loop
    g_main_loop_run (mainloop);

    // stop recording
    if (log_fname && logger_unit)
        cam_unit_set_control_boolean (logger_unit, "record", FALSE);

    // cleanup
    status = 0;
done:
    if (mainloop) g_main_loop_unref (mainloop);
    if (chain) {
        cam_unit_chain_all_units_stream_shutdown (chain);
        g_object_unref (chain);
    }
    free(input_id);
    free(log_fname);
    free(chain_fname);
    if (self) {
        free (self);
    }
    return status;
}
